opencl-c-mode.el

```diff
@@ -0,0 +1,808 @@
+If any of these also are on `c-type-list-kwds', `c-ref-list-kwds',
+`c-colon-type-list-kwds', `c-paren-nontype-kwds', `c-paren-type-kwds',
+`c-<>-type-kwds', or `c-<>-arglist-kwds' then the associated clauses
+will be handled.
+
+Do not try to modify this list for end user customizations; the
+`*-font-lock-extra-types' variable, where `*' is the mode prefix, is
+the appropriate place for that."
+  glsl
+  (append
+   opencl-c-primitive-types
+   opencl-c-atomic-types
+   opencl-c-vector-types
+   opencl-c-descriptor-types
+   ;; Use append to not be destructive on the return value below.
+   (append
+    (c-lang-const c-primitive-type-kwds)
+    nil)))
+
+(c-lang-defconst c-modifier-kwds
+  opencl-c
+  (append (c-lang-const c-modifier-kwds)
+          '("__kernel" "__global" "__local" "__constant" "__private" "__generic"
+            "__read_only" "__write_only" "__read_write"
+            "kernel" "global" "local" "constant" "private" "generic"
+            "read_only" "write_only" "read_write"
+            "uniform" "pipe")))
+
+(defconst opencl-c-font-lock-keywords-1 (c-lang-const c-matchers-1 opencl-c))
+(defconst opencl-c-font-lock-keywords-2 (c-lang-const c-matchers-2 opencl-c))
+(defconst opencl-c-font-lock-keywords-3 (c-lang-const c-matchers-3 opencl-c))
+(defvar opencl-c-font-lock-keywords (c-lang-const c-matchers-3 opencl-c))
+(defun opencl-c-font-lock-keywords ()
+  "Compose a list of font-locking keywords for `opencl-c-mode'."
+  (c-compose-keywords-list opencl-c-font-lock-keywords))
+
+
+;;;###autoload
+(progn
+  (add-to-list 'auto-mode-alist '("\\.cl\\'" . opencl-c-mode))
+  (add-to-list 'auto-mode-alist '("\\.clc\\'" . opencl-c-mode))
+  (add-to-list 'auto-mode-alist '("\\.opencl\\'" . opencl-c-mode)))
+
+;;;###autoload
+(define-derived-mode opencl-c-mode prog-mode "OpenCL[C]"
+  "Major mode for editing OpenCL C kernels.
+
+\\{opencl-c-mode-map}"
+  :syntax-table opencl-c-mode-syntax-table
+  :after-hook (progn (c-make-noise-macro-regexps)
+		     (c-make-macro-with-semi-re)
+		     (c-update-modeline))
+  (c-initialize-cc-mode t)
+  (c-init-language-vars opencl-c-mode)
+  (c-common-init 'opencl-c-mode)
+  (cc-imenu-init cc-imenu-c++-generic-expression)
+
+  (c-run-mode-hooks 'c-mode-common-hook)
+  (run-mode-hooks 'opencl-c-mode-hook)
+
+  (setq-local comment-start "// ")
+  (setq-local comment-end "")
+  (setq-local comment-padding "")
+
+  (font-lock-add-keywords nil opencl-c-font-lock-keywords))
+
+(defun opencl-c-lookup ()
+  "Get OpenCL documentation for string in region or point."
+  (interactive)
+  (let* ((api-function (if (region-active-p)
+                           (buffer-substring-no-properties (region-beginning)
+                                                           (region-end))
+                         (thing-at-point 'symbol)))
+         (doc-url (concat
+                   "http://www.khronos.org/registry/cl/sdk/2.1/docs/man/xhtml/"
+                   api-function ".html")))
+    (browse-url doc-url)))
+
```

Comment:
Let's also add  `(define-key opencl-mode-map (kbd "C-c ! d") 'opencl-lookup)`  for the mapping